src/blocks/ibus.rs

```diff
@@ -0,0 +1,185 @@
+        thread::spawn(move || {
+            let c = Connection::open_private(&ibus_address).unwrap();
+            c.add_match(
+                "interface='org.freedesktop.IBus',member='GlobalEngineChanged'",
+            ).unwrap();
+            loop {
+                for ci in c.iter(100000) {
+                    if let Some(engine_name) = parse_msg(&ci) {
+                            let mut engine = engine_original.lock().unwrap();
+                            *engine = engine_name.to_string();
+                            // Tell block to update now.
+                            send.send(Task {
+                                id: id.clone(),
+                                update_time: Instant::now(),
+                            });
+                    };
+                }
+            }
+        });
+
+        Ok(IBus {
+            id: id_copy,
+            text: TextWidget::new(config.clone()).with_text("IBus"),
+            engine
+        })
+    }
+}
+
+impl Block for IBus {
+    fn id(&self) -> &str {
+        &self.id
+    }
+
+    // Updates the internal state of the block.
+    fn update(&mut self) -> Result<Option<Duration>> {
+        let string = (*self.engine
+            .lock()
+            .block_error("ibus", "failed to acquire lock")?)
+            .clone();
+        self.text.set_text(string);
+        Ok(None)
+    }
+
+    // Returns the view of the block, comprised of widgets.
+    fn view(&self) -> Vec<&I3BarWidget> {
+        vec![&self.text]
+    }
+
+    // This function is called on every block for every click.
+    // TODO: Filter events by using the event.name property,
+    // and use to switch between input engines?
+    fn click(&mut self, _: &I3BarEvent) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn parse_msg(ci: &ConnectionItem) -> Option<&str> {
+    let m = if let &ConnectionItem::Signal(ref s) = ci { s } else { return None };
+    if &*m.interface().unwrap() != "org.freedesktop.IBus" { return None };
+    if &*m.member().unwrap() != "GlobalEngineChanged" { return None };
+    let engine = m.get1::<&str>();
+    engine
+}
+
+// Gets the address being used by the currently running ibus daemon.
+// 
+// By default ibus will write the address to `$XDG_CONFIG_HOME/ibus/bus/aaa-bbb-ccc`
+// where aaa = dbus machine id, usually found at /etc/machine-id
+//       bbb = hostname - seems to be "unix" in most cases [see L99 of reference]
+//       ccc = display number from $DISPLAY
+// Refer to: https://github.com/ibus/ibus/blob/7cef5bf572596361bc502e8fa917569676a80372/src/ibusshare.c
+//
+// Example file contents:
+// ```
+// # This file is created by ibus-daemon, please do not modify it
+// IBUS_ADDRESS=unix:abstract=/tmp/dbus-8EeieDfT,guid=7542d73dce451c2461a044e24bc131f4
+// IBUS_DAEMON_PID=11140
+// ```
+fn get_ibus_address() -> Result<String> {
+    // TODO: Check IBUS_ADDRESS variable, as it seems it can be manually set too.
+
+    let config_dir = env::var("XDG_CONFIG_HOME")
```

Comment:
Do I understand correctly that this will error out if this variable is unset? Or is it just for errors in actually retrieving the environment variable?

Reply:
> The returned result is Ok(s) if the environment variable is present and is valid unicode. If the environment variable is not present, or it is not valid unicode, then Err will be returned.

Errors if not set. I guess I should update the block_error message to make it clearer.

Reply:
You should also fall back on the default specified by the XDG standard, which is `~/.config` I believe.

Reply:
Agreed: 
> If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.